### DISCUSSION TITLE
Maintain field order when rendering templates

### DISCUSSION THREAD

**Discussion by verdverm** [cue-lang/cue#2115] [Q&A]
I suspect this is tied to Decode and maybe Go maps? At some point the fields became lexically ordered.

It would be very helpful if fields maintained their order, like export

Given:

```
import "text/template"

foo: {
	b: "b"
	a: "a"
}

tpl: """
{{ range $k, $v := . }}
{{ $k }}: {{ $v }}
{{ end }}
"""

txt: template.Execute(tpl, foo)
```

Output:

```
$ cue export test.cue -e foo
{
    "b": "b",
    "a": "a"
}
```
note how a & b swap order in this example
```
$ cue eval test.cue 
foo: {
    b: "b"
    a: "a"
}
tpl: """
    {{ range $k, $v := . }}
    {{ $k }}: {{ $v }}
    {{ end }}
    """
txt: """

    a: a

    b: b

    """
```

**Comment by myitcv** [comment]
Thanks for raising this, @verdverm.

If you think this is a bug, please can you update the issue to follow the ["Bug report" template](https://github.com/cue-lang/cue/issues/new/choose)?

Thanks

**Comment by verdverm** [comment]
I'm not sure if it is a bug, or even something CUE can handle easily. It needs more investigation to find the root cause (Decode implementation or Go map iteration)

**Reply by myitcv** [reply]
I've converted this to a discussion until we work out concretely what the next steps are.

Per the [docs for `range` in `text/template`]():

```
	If the value is a map and the
	keys are of basic type with a defined order, the elements will be
	visited in sorted key order.
```

So this appears to be working as intended from a `text/template` perspective.

However, I can see an argument in the context of #474 that we should retain the CUE-defined order. What sort of API surface that would require I'm not clear.
